library/dhm.c

```diff
@@ -198,19 +198,18 @@ int mbedtls_dhm_make_params( mbedtls_dhm_context *ctx, int x_size,
 
     if( mbedtls_mpi_cmp_int( &ctx->P, 0 ) == 0 )
         return( MBEDTLS_ERR_DHM_BAD_INPUT_DATA );
+    n_bits = mbedtls_mpi_bitlen( &ctx->P );
 
     /*
-     * Generate X as large as possible ( < P )
+     * Generate X as large as possible ( 2 <= X <= P - 2 )
      */
     do
     {
-        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
-
-        while( mbedtls_mpi_cmp_mpi( &ctx->X, &ctx->P ) >= 0 )
-            MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 1 ) );
-
-        if( count++ > 10 )
+        if( count++ > 30 )
             return( MBEDTLS_ERR_DHM_MAKE_PARAMS_FAILED );
+
+        MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_shift_r( &ctx->X, 8 * x_size - n_bits ) );
```

Comment:
This is correct when `x_size` is the size of `P`. But when `x_size` is smaller, it's reducing the range. (`x_size` larger than the size of `P` doesn't make sense, so it's ok if the behavior changes in that case, although in an LTS branch we should be conservative even in error cases.)

I think this would be a correct strategy:
```
if( 8 * x_size < n_bits )
    MBEDTLS_MPI_CHK( mbedtls_mpi_fill_random( &ctx->X, x_size, f_rng, p_rng ) );
else
{
    // the loop, with 8 * n_bits instead of x_size
}
```